src/cdomains/arrayDomain.ml

```diff
@@ -353,15 +353,19 @@ struct
                 begin
                   match Idx.to_int l with
                   | Some i ->
-                    let b = ask.f (Q.MayBeLess (exp, Cil.kintegerCilint (Cilfacade.ptrdiff_ikind ()) i)) in
+                    let b = Q.may_be_less ask exp (Cil.kintegerCilint (Cilfacade.ptrdiff_ikind ()) i) in
                     not b (* !(e <_{may} length) => e >=_{must} length *)
                   | None -> false
                 end
               | _ -> false
             in
             let e_must_less_zero =
-              let b = ask.f (Q.MayBeLess (Cil.mone, exp)) in
-              not b (* !(-1 <_{may} e) => e <=_{must} -1 *)
+              if Cil.isSigned (Cilfacade.get_ikind_exp exp) then ( (* avoid may_be_less comparing unsigned index to -1, which is promoted to max int of ikind via unsigned underflow *)
```

Comment:
>avoid may_be_less comparing unsigned index to -1, which is promoted to max int of ikind via unsigned underflow

Is this something that one has to take additional care of now, when rephrasing the `MayBeLess` (and related) queries to `EvalInt`? Although it is not so clear what happens in such cases in the current implementation, as the int-domains also consider the `ikind`s. So it probably should be implemented like that here, anyways.